File: reverse_kozai_lidov.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Apr 14 15:58:31 2026

@author: resulayberksahpaz
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp
from matplotlib.animation import FuncAnimation

# --- Physical Constants (SI) ---
G = 6.67430e-11
MSUN = 1.989e30
MJUP = 1.898e27
RJUP = 7.149e7
AU = 1.496e11
C = 2.9979e8
YEAR_SEC = 31557600

def secular_dynamics_reverse(t, y, m1, m2, m3, a_out, e_out, tv2_yr):
    a, e, inc, omega = y
    
    # Boundary clamping for eccentricity
    e = max(min(e, 0.9999), 1e-6)
    
    # Kozai-Lidov Timescale
    p_in = 2 * np.pi * np.sqrt(a**3 / (G * (m1 + m2)))
    p_out = 2 * np.pi * np.sqrt(a_out**3 / (G * (m1 + m2 + m3)))
    tau_kl = (2 * p_out**2 / (3 * np.pi * p_in)) * ((m1 + m2 + m3) / m3) * (1 - e_out**2)**1.5
    
    # KL Dynamics
    de_dt_kl = (15/8) * (e * np.sqrt(1 - e**2) / tau_kl) * np.sin(inc)**2 * np.sin(2 * omega)
    di_dt_kl = -(15/8) * (e**2 / (tau_kl * np.sqrt(1 - e**2))) * np.sin(inc) * np.cos(inc) * np.sin(2 * omega)
    domega_dt_kl = (3 / (4 * tau_kl * np.sqrt(1 - e**2))) * (2 * (1 - e**2) + 5 * np.sin(omega)**2 * (e**2 - np.sin(inc)**2))
    
    # GR Precession
    n = np.sqrt(G * (m1 + m2) / a**3)
    domega_dt_gr = (3 * G * (m1 + m2) * n) / (a * C**2 * (1 - e**2))
    
    # Tidal Migration (Anti-dissipation logic)
    tv_sec = tv2_yr * YEAR_SEC
    tf2 = tv_sec * (9/8) * (a / RJUP)**6 * (m1 / m2)
    
    v2 = (9 / tf2) * (1 + (15/4)*e**2 + (15/8)*e**4 + (5/64)*e**6) / (1 - e**2)**6.5
    w2 = (1 / tf2) * (1 + (15/2)*e**2 + (45/8)*e**4 + (5/16)*e**6) / (1 - e**2)**6.5
    
    da_dt_tide = -a * (2 * v2 * (e**2 / (1 - e**2)) + 2 * w2)
    de_dt_tide = -e * (v2 + 9 * w2)
    
    return [da_dt_tide, de_dt_kl + de_dt_tide, di_dt_kl, domega_dt_kl + domega_dt_gr]

def run_simulation():
    # --- System Setup ---
    m1, m2, m3 = 0.6 * MSUN, 0.390 * MJUP, 0.31 * MSUN
    a_out, e_out, tv2 = 80 * AU, 0.2, 0.1
    
    # Current state
    p_days = 10
    target_p_sec = p_days * 24 * 3600
    a_start = ((target_p_sec * np.sqrt(G * (m1 + m2))) / (2 * np.pi))**(2/3)
    y0 = [a_start, 0.01, np.radians(85), np.radians(90)]
    
    # Integration range (0 to -100 Myr)
    t_span = (0, -100e12 * YEAR_SEC)
    a_birth_limit = 5.0 * AU
    
    # Stop event: when a reaches the birth limit
    def stop_at_birth(t, y, *args):
        return y[0] - a_birth_limit
    stop_at_birth.terminal = True
    stop_at_birth.direction = 1 # 'a' increases moving backwards
    
    print("Running Time-Reversal Integration...")
    sol = solve_ivp(secular_dynamics_reverse, t_span, y0, 
                    args=(m1, m2, m3, a_out, e_out, tv2),
                    events=stop_at_birth, rtol=1e-8, atol=1e-10)
    
    # Print Stats
    t_myr = np.abs(sol.t / (1e6 * YEAR_SEC))
    print(f"Simulation went back to: {t_myr[-1]:.2f} Myr ago")
    print(f"Original Semi-major Axis: {sol.y[0,-1]/AU:.4f} AU")
    print(f"Original Eccentricity: {sol.y[1,-1]:.4f}")
    
    visualize(sol, m1, m3, a_out, e_out)

def get_orbit_path(a, e, inc, omega):
    theta = np.linspace(0, 2 * np.pi, 100)
    r = a * (1 - e**2) / (1 + e * np.cos(theta))
    x_orb = r * np.cos(theta)
    y_orb = r * np.sin(theta)
    
    # Rotations
    x = x_orb * np.cos(omega) - y_orb * np.sin(omega)
    y_temp = x_orb * np.sin(omega) + y_orb * np.cos(omega)
    y = y_temp * np.cos(inc)
    z = y_temp * np.sin(inc)
    return x, y, z

def visualize(sol, m1, m3, a_out, e_out):
    t_myr = np.abs(sol.t / (1e6 * YEAR_SEC))
    y = sol.y
    p_out = 2 * np.pi * np.sqrt(a_out**3 / (G * (m1 + m3)))
    
    fig = plt.figure(figsize=(14, 8))
    gs = fig.add_gridspec(3, 2)
    
    ax_a = fig.add_subplot(gs[0, 0])
    ax_e = fig.add_subplot(gs[1, 0])
    ax_s = fig.add_subplot(gs[2, 0])
    ax_3d = fig.add_subplot(gs[:, 1], projection='3d')
    
    # Plot static lines
    ax_a.plot(t_myr, y[0]/AU, 'b'); ax_a.set_ylabel('a (AU)'); ax_a.grid(True)
    ax_e.plot(t_myr, y[1], 'r'); ax_e.set_ylabel('e'); ax_e.grid(True)
    
    r_sep = a_out * (1 - e_out * np.cos(2 * np.pi * sol.t / p_out))
    ax_s.plot(t_myr, r_sep/AU, color='gray'); ax_s.set_ylabel('r_sep (AU)'); ax_s.set_xlabel('Myr Ago'); ax_s.grid(True)
    
    # 3D setup
    a1, a3 = a_out * (m3/(m1+m3)), a_out * (m1/(m1+m3))
    limit = (a3/AU) * 1.2
    
    h_star1, = ax_3d.plot([0], [0], [0], 'yo', markersize=10)
    h_star3, = ax_3d.plot([0], [0], [0], 'ko', markersize=8)
    h_orbit, = ax_3d.plot([], [], [], 'k-', alpha=0.5)
    h_planet, = ax_3d.plot([], [], [], 'ro')
    
    ax_3d.set_xlim([-limit, limit]); ax_3d.set_ylim([-limit, limit]); ax_3d.set_zlim([-limit/2, limit/2])
    
    def update(i):
        idx = i * (len(sol.t) // 200) # Speed up animation
        if idx >= len(sol.t): idx = len(sol.t) - 1
        
        phase = 2 * np.pi * (sol.t[idx] / p_out)
        r_f = (1 - e_out * np.cos(phase))
        x1, y1 = -a1 * r_f * np.cos(phase), -a1 * r_f * np.sin(phase)
        x3, y3 =  a3 * r_f * np.cos(phase),  a3 * r_f * np.sin(phase)
        
        h_star1.set_data([x1/AU], [y1/AU]); h_star1.set_3d_properties([0])
        h_star3.set_data([x3/AU], [y3/AU]); h_star3.set_3d_properties([0])
        
        ox, oy, oz = get_orbit_path(y[0,idx], y[1,idx], y[2,idx], y[3,idx])
        h_orbit.set_data((ox+x1)/AU, (oy+y1)/AU); h_orbit.set_3d_properties(oz/AU)
        
        return h_star1, h_star3, h_orbit

    # In Python, we usually just show the final plot or save an animation
    plt.tight_layout()
    plt.show()

if __name__ == "__main__":
    run_simulation()

File: reverse_kozai_lidov.m
function KL_Reverse_Migration_Sim()
    % --- Physical Constants (SI) ---
    G = 6.67430e-11; 
    MSun = 1.989e30; 
    MJup = 1.898e27; 
    RJup = 7.149e7;  
    AU = 1.496e11;   
    c = 2.9979e8;    
    
    % --- "Current" System State (Starting Point for Backwards) ---
    m1 = 0.6 * MSun;   
    m2 = 0.390 * MJup;   
    m3 = 0.31 * MSun;   
    
    % Assuming we found a planet with a 10-day period
    current_P_days = 10;
    target_P_sec = current_P_days * 24 * 3600;
    a_current = ((target_P_sec * sqrt(G*(m1+m2))) / (2*pi))^(2/3);
    
    e_current = 0.01;            % Circularized orbit
    inc_current = deg2rad(85);   % Current inclination (relative to binary)
    omega_current = deg2rad(90); % Argument of periastron
    
    a_out = 80 * AU;  
    e_out = 0.2; 
    tV2 = 0.1; 
    
    % --- Reverse Simulation Setup ---
    % t_span goes from 0 to -100 Myr
    t_end_backward = -100e8 * 31557600; 
    y_current = [a_current, e_current, inc_current, omega_current];
    
    % Stop if the planet reaches a likely birth semi-major axis (e.g., 5 AU)
    a_birth_limit = 5.0 * AU;
    options = odeset('RelTol', 1e-8, 'AbsTol', 1e-10, ...
                     'Events', @(t,y) stop_at_birth(t,y,a_birth_limit));
    
    fprintf('Running Time-Reversal Integration...\n');
    [t, y] = ode45(@(t, y) secular_dynamics_REVERSE(t, y, m1, m2, m3, a_out, e_out, tV2, RJup, G, c), ...
                   [0, t_end_backward], y_current, options);
               
    % --- Final Statistics (The "Past" State) ---
    t_myr_final = t(end)/(1e6 * 31557600);
    fprintf('Simulation went back to: %.2f Myr ago\n', abs(t_myr_final));
    fprintf('Likely original Semi-major Axis: %.4f AU\n', y(end,1)/AU);
    fprintf('Likely original Eccentricity: %.4f\n', y(end,2));
    fprintf('Likely original Inclination: %.2f degrees\n', rad2deg(y(end,3)));

    % --- Animated Visualization ---
    visualize_KL_animation(t, y, m1, m3, a_out, e_out, AU, G);
end

% --- Stop when the planet migrates "out" to its starting distance ---
function [value, isterminal, direction] = stop_at_birth(~, y, a_limit)
    value = y(1) - a_limit; 
    isterminal = 1;          
    direction = 1; % Trigger when 'a' is increasing (moving backwards)
end

function dydt = secular_dynamics_REVERSE(~, y, m1, m2, m3, a_out, e_out, tV2, RJup, G, c)
    a = y(1); e = y(2); inc = y(3); omega = y(4);
    
    % Keep eccentricity in physical bounds
    e = max(min(e, 0.9999), 1e-6); 
    
    Pin = 2*pi*sqrt(a^3 / (G*(m1+m2)));
    Pout = 2*pi*sqrt(a_out^3 / (G*(m1+m2+m3)));
    tau_KL = (2*Pout^2 / (3*pi*Pin)) * ((m1+m2+m3)/m3) * (1-e_out^2)^1.5;
    
    % Reversible Secular Dynamics (KL + GR)
    de_dt_KL = (15/8) * (e * sqrt(1-e^2) / tau_KL) * sin(inc)^2 * sin(2*omega);
    di_dt_KL = -(15/8) * (e^2 / (tau_KL * sqrt(1-e^2))) * sin(inc) * cos(inc) * sin(2*omega);
    domega_dt_KL = (3/(4*tau_KL*sqrt(1-e^2))) * (2*(1-e^2) + 5*sin(omega)^2*(e^2-sin(inc)^2));
    
    n = sqrt(G*(m1+m2)/a^3);
    domega_dt_GR = (3 * G * (m1 + m2) * n) / (a * c^2 * (1 - e^2));
    
    % Tidal Migration (Will be handled as "anti-dissipation" by the negative t_span)
    tV_sec = tV2 * 31557600; 
    tF2 = tV_sec * (9/8) * (a/RJup)^6 * (m1/m2); 
    
    V2 = (9/tF2) * (1 + (15/4)*e^2 + (15/8)*e^4 + (5/64)*e^6) / (1-e^2)^6.5;
    W2 = (1/tF2) * (1 + (15/2)*e^2 + (45/8)*e^4 + (5/16)*e^6) / (1-e^2)^6.5;
    
    da_dt_tide = -a * (2*V2 * (e^2/(1-e^2)) + 2*W2); 
    de_dt_tide = -e * (V2 + 9*W2); 
    
    dydt = [da_dt_tide; de_dt_KL + de_dt_tide; di_dt_KL; domega_dt_KL + domega_dt_GR];
end

function visualize_KL_animation(t, y, m1, m3, a_out, e_out, AU, G)
    % Use absolute value for time to plot 0 -> Past
    t_myr = abs(t / (1e6 * 31557600)); 
    fig = figure('Color', 'w', 'Position', [50, 50, 1300, 700]);
    
    subplot(3, 2, 1);
    plot(t_myr, y(:,1)/AU, 'b', 'LineWidth', 1.5);
    ylabel('a (AU)'); title('Semi-major Axis (Backward)'); grid on;
    h_a_mark = line(t_myr(1), y(1,1)/AU, 'Marker', 'o', 'Color', 'r', 'MarkerFaceColor', 'r');

    subplot(3, 2, 3);
    plot(t_myr, y(:,2), 'r', 'LineWidth', 1);
    ylabel('e'); title('Eccentricity (Backward)'); grid on;
    h_e_mark = line(t_myr(1), y(1,2), 'Marker', 'o', 'Color', 'b', 'MarkerFaceColor', 'b');
    
    subplot(3, 2, 5);
    P_out = 2*pi*sqrt(a_out^3 / (G*(m1+m3)));
    r_sep = a_out * (1 - e_out * cos(2 * pi * t / P_out));
    plot(t_myr, r_sep/AU, 'Color', [0.4 0.4 0.4], 'LineWidth', 1.5);
    ylabel('r_{sep} (AU)'); xlabel('Myr Ago'); title('Stellar Separation'); grid on;
    h_r_mark = line(t_myr(1), r_sep(1)/AU, 'Marker', 'o', 'Color', 'k', 'MarkerFaceColor', 'k');

    ax3 = subplot(3, 2, [2, 4, 6]);
    hold on; grid on; axis equal;
    
    a1 = a_out * (m3 / (m1 + m3)); 
    a3 = a_out * (m1 / (m1 + m3)); 
    h_star1  = plot3(0, 0, 0, 'yo', 'MarkerSize', 10, 'MarkerFaceColor', 'y', 'DisplayName', 'Host Star');
    h_star3  = plot3(0, 0, 0, 'ko', 'MarkerSize', 8, 'MarkerFaceColor', [0.5 0.5 0.5], 'DisplayName', 'Companion');
    h_orbit  = plot3(0, 0, 0, 'k-', 'LineWidth', 1.2); 
    h_planet = plot3(0, 0, 0, 'ro', 'MarkerSize', 6, 'MarkerFaceColor', 'r'); 
    
    xlabel('X (AU)'); ylabel('Y (AU)'); zlabel('Z (AU)');
    view(45, 30);
    limit = (a3/AU) * 1.5;
    axis(ax3, [-limit limit -limit limit -limit/2 limit/2]);

    skip = max(1, round(length(t)/400)); 
    for i = 1:skip:length(t)
        if ~ishandle(fig), break; end
        phase = 2 * pi * (t(i) / P_out);
        r_fact = (1 - e_out * cos(phase)); 
        x1 = -a1 * r_fact * cos(phase); y1 = -a1 * r_fact * sin(phase);
        x3 =  a3 * r_fact * cos(phase); y3 =  a3 * r_fact * sin(phase);
        
        set(h_star1, 'XData', x1/AU, 'YData', y1/AU);
        set(h_star3, 'XData', x3/AU, 'YData', y3/AU);
        
        [X, Y, Z] = get_orbit_path(y(i,1), y(i,2), y(i,3), y(i,4));
        set(h_orbit, 'XData', (X+x1)/AU, 'YData', (Y+y1)/AU, 'ZData', Z/AU);
        
        r_p = (y(i,1)*(1-y(i,2)));
        xp_rel = r_p * cos(y(i,4));
        yp_temp_rel = r_p * sin(y(i,4));
        set(h_planet, 'XData', (xp_rel+x1)/AU, ...
                      'YData', (yp_temp_rel*cos(y(i,3))+y1)/AU, ...
                      'ZData', (yp_temp_rel*sin(y(i,3)))/AU);
        
        set(h_a_mark, 'XData', t_myr(i), 'YData', y(i,1)/AU);
        set(h_e_mark, 'XData', t_myr(i), 'YData', y(i,2));
        set(h_r_mark, 'XData', t_myr(i), 'YData', r_sep(i)/AU);
        drawnow; 
    end
end

function [x, y, z] = get_orbit_path(a, e, inc, omega)
    theta = linspace(0, 2*pi, 100);
    r = a * (1 - e^2) ./ (1 + e * cos(theta));
    x_orb = r .* cos(theta);
    y_orb = r .* sin(theta);
    x = x_orb * cos(omega) - y_orb * sin(omega);
    y_temp = x_orb * sin(omega) + y_orb * cos(omega);
    y = y_temp * cos(inc);
    z = y_temp * sin(inc);
end